### ISSUE TITLE
Deadlock on PurchaseAsync (Droid)

### ISSUE THREAD

**Issue by Sebastian1989101** [jamesmontemagno/InAppBillingPlugin#92]
If you are creating an issue for a BUG please fill out this information. If you are asking a question or requesting a feature you can delete the sections below. 

**Failure to fill out this information will result in this issue being closed.** If you post a full stack trace in a bug it will be closed, please post it to http://gist.github.com and then post the link here.

## Bug Information

Version Number of Plugin: 1.2.3.107
Device Tested On: Multiple Android Devices 
Simulator Tested On: -
Version of VS: 15.4.2
Version of Xamarin: 4.7.10.33
Versions of other things you are using: -

### Steps to reproduce the Behavior
Use a Android device without an attached Google account (I guess it would be the same on a Simulator with installed PlayStore but without authentication but I'm not 100% sure about that). 

### Expected Behavior 
No deadlock / a InAppBillingPurchaseException

### Actual Behavior 
Deadlock on call of "await CrossInAppBilling.Current.PurchaseAsync" while the device shows the error message "Authentication is required. You need to sign into your Google Account." (this is not shown by the App it is a system dialog). 

### Code snippet
```
       public async Task PurchaseAsync(string identifier)
        {
            try
            {
                var connection = await CrossInAppBilling.Current.ConnectAsync();
                if (!connection)
                {
                    _messenger.Publish(new InAppPurchaseResponse(this, false, AppResources.InAppPurchase_ErrorConnectingStore));
                    return;
                }

                var purchase = await CrossInAppBilling.Current.PurchaseAsync(identifier, ItemType.InAppPurchase, "apppayload", _verifyPurchase); //sometimes deadlock?!
                if (purchase == null)
                {
                    _messenger.Publish(new InAppPurchaseResponse(this, false, AppResources.InAppPurchase_ErrorPurchasingProduct));
                    return;
                }

                _messenger.Publish(MarkProductAsPurchased(identifier) ? new InAppPurchaseResponse(this, true, AppResources.InAppPurchase_PurchaseSuccessful) : new InAppPurchaseResponse(this, false, AppResources.InAppPurchase_ErrorPurchaseActivation));
            }
            catch (InAppBillingPurchaseException ex)
            {
                _messenger.Publish(new InAppPurchaseResponse(this, false, ex.Message));
            }
            finally
            {
                await CrossInAppBilling.Current.DisconnectAsync();
            }
        }
```

### Screenshotst
https://i.imgur.com/QP5DEUr.png

**Comment by jamesmontemagno** [comment]
You may have to debug through this one as I don't have a device currently for testing this situation currently and out of the country.

**Comment by jamesmontemagno** [comment]
I think this may be fixed with #111  perhaps.

**Comment by AndreiMisiukevich** [comment]
Hi
I haven't reproduced this issue on my phone (Samsung Galaxy S7).

Because if I have no google account, i get FALSE from ConnectAsync method..

**Comment by Sebastian1989101** [comment]
More and more users reporting they don't get the activation of the products within the app on Android. It seems like there is a deadlock somewhere because the Analytics.TrackEvent calls do not match. For the latest version I got 1079 entries for "Purchase unlock started" but only 319 for "Purchase unlock finished" and onlz 371 for "Disconnect after purchase unlock". After a restore they all get the unlocks but because this happens in a deadlock like style, it never activates. 

I even added this because my first thought was that Google needs to long to verify the process, but after I checked the logging, I'm pretty sure that a deadlock happens somewhere. 
```
if (Device.RuntimePlatform == Device.Android)
{
	await Task.Delay(1000);
	await RestoreUnlockAsync(true);
}
```

```
public async Task PurchaseUnlockAsync(string identifier)
{
	try
	{
		Analytics.TrackEvent("Purchase unlock started", new Dictionary<string, string> {{nameof(identifier), identifier}});

		var connection = await InAppBillingPlatformWrapper.Implementation.ConnectAsync();
		if (!connection)
		{
			_messenger.Publish(new InAppPurchaseResponse(this, false, AppResources.InAppPurchase_ErrorConnectingStore));
			return;
		}

		var purchase = await InAppBillingPlatformWrapper.Implementation.PurchaseAsync(identifier, ItemType.InAppPurchase, "apppayload", _verifyPurchase);
		if (purchase == null)
		{
			_messenger.Publish(new InAppPurchaseResponse(this, false, AppResources.InAppPurchase_ErrorPurchasingProduct));
			return;
		}

		var success = MarkProductAsPurchased(identifier);
		_messenger.Publish(success ? new InAppPurchaseResponse(this, true, AppResources.InAppPurchase_PurchaseSuccessful) : new InAppPurchaseResponse(this, false, AppResources.InAppPurchase_ErrorPurchaseActivation));
		Analytics.TrackEvent("Purchase unlock finished", new Dictionary<string, string> {{nameof(identifier), identifier}, {"success", success.ToString()}});

		if (Device.RuntimePlatform == Device.Android)
		{
			await Task.Delay(1000);
			await RestoreUnlockAsync(true);
		}
	}
	catch (InAppBillingPurchaseException ex)
	{
		_messenger.Publish(new InAppPurchaseResponse(this, false, ex.Message));
		Crashes.TrackError(ex, new Dictionary<string, string> { { "version", _versionService.GetAppVersion() } });
	}
	catch (Exception ex)
	{
		Crashes.TrackError(ex, new Dictionary<string, string> { { "version", _versionService.GetAppVersion() } });
		throw;
	}
	finally
	{
		Analytics.TrackEvent("Disconnect after purchase unlock", new Dictionary<string, string> {{nameof(identifier), identifier}});
		await InAppBillingPlatformWrapper.Implementation.DisconnectAsync();
	}
}
```

So maybe this is the same or a similar problem to this issue here. It does not happen on my devices yet so it's kinda hard to debug it.

**Comment by theD99** [comment]
I'm getting this as well.

Calling PurchaseAsync seems to work to a point... the purchase seems to get to Google as I can call GetPurchasesAsync which returns the purchase receipt that I'd expect, but the actual PurchaseAsync call never comes back.

**Comment by jamesmontemagno** [comment]
This is probably some case when they don't have a google account setup, probably pretty rare and can be tested maybe on emulators? There were some enhancements to the code in the latest betas around there. If you can reproduce it locally can you debug through the code and see what is happening.

**Comment by Sebastian1989101** [comment]
@theD99 what did you do to reproduce the issue? I would like to debug it too for my apps but none of my devices will create this issue :(

**Comment by theD99** [comment]
I pretty much swapped out my entire implementation with the source code and noted that no matter what I did, the TaskCompletionSource at the bottom of PurchaseAsync was never coming back which, obviously, was because SetResult wasn't being called on it from my main activity (I was jumping the thread before I got there--dumb mistake). 

So I'd start with OnActivityResult and go from there.

**Comment by theD99** [comment]
Yeah, SetResult is called on the Task in HandleActivityResult which is what triggers PurchaseAsync to come back.

**Comment by Sebastian1989101** [comment]
hmm seems like it was a threading problem for me too at this point. Didn't even notice that I change the code there for some reason. So I guess this only happens with a missing Google Account and/or if the threading is messed up for the SetResult?

**Comment by EmilAlipiev** [comment]
Hi, just off the topic.  Will you provide a PR for Amazon in app billing or at least can you share your implementation on a separate issue maybe?

**Comment by Sebastian1989101** [comment]
@EmilAlipiev Sure I can share the implementation once it is finished. But I can only start with the implementation if Amazon fixes this bug here: https://forums.developer.amazon.com/questions/115917/iap-in-xamarin-problem-with-amazoniapv2bridge-libr.html (until now, they just ignored it - I guess they think there is no need for a working Xamarin version of theire API...). 

@jamesmontemagno I guess this issue can be closed. This problem only happens if the Google Account is missing or if the `OnActivityResult` is messed up.

**Comment by EmilAlipiev** [comment]
@Sebastian1989101 oh, this is really annoying how hard can this be to fix but they dont care. I wish google purchases are possible on amazon store. I wonder if they have policy against external payment methods if I just want to integrate my own one.

**Comment by EmilAlipiev** [comment]
@Sebastian1989101  did you find any solution for amazon? I am searching for api end points. do you know if it is simply possible using http post request using HttpClient

**Comment by EmilAlipiev** [comment]
@Sebastian1989101 I found this recently. https://stackoverflow.com/questions/45554652/amazon-iap-not-calling-listener-call-backs-kindle-fire-xamarin
this guy claims to achieve using Android library with the java bindings. Have you tried that?

**Comment by EmilAlipiev** [comment]
@Sebastian1989101  i just come back to amazon issue and actually latest xamarin dll works for me. it doesnt throw exception. i have tested pulling the subscriptions using amazon iap tester and it was all fine :)
